libp2p/transports/tls/certificate.nim

```diff
@@ -0,0 +1,489 @@
+  CertificateCreationError* = object of TLSCertificateError
+  CertificateParsingError* = object of TLSCertificateError
+  IdentityPubKeySerializationError* = object of TLSCertificateError
+  IdentitySigningError* = object of TLSCertificateError
+
+# Define the P2pExtension and P2pCertificate types
+type
+  P2pExtension* = object
+    publicKey*: seq[byte]
+    signature*: seq[byte]
+
+  P2pCertificate* = object
+    certificate*: mbedtls_x509_crt
+    extension*: P2pExtension
+
+proc ptrInc*(p: ptr byte, n: uint): ptr byte =
+  ## Utility function to increment a pointer by n bytes.
+  cast[ptr byte](cast[uint](p) + n)
+
+proc generateSignedKey(
+    signature: seq[byte], pubKey: seq[byte]
+): seq[byte] {.raises: [ASN1EncodingError].} =
+  ## Generates the ASN.1-encoded SignedKey structure.
+  ##
+  ## The SignedKey structure contains the public key and its signature,
+  ## encoded as a SEQUENCE of two OCTET STRINGs.
+  ##
+  ## Parameters:
+  ## - `signature`: The signature bytes.
+  ## - `pubKey`: The public key bytes.
+  ##
+  ## Returns:
+  ## A sequence of bytes representing the ASN.1-encoded SignedKey.
+  ##
+  ## Raises:
+  ## - `ASN1EncodingError` if ASN.1 encoding fails.
+  const extValueSize = 256 # Buffer size for ASN.1 encoding
+  var
+    extValue: array[extValueSize, byte]
+    extPtr: ptr byte = addr extValue[extValueSize - 1] # Start at the end of the buffer
+    startPtr: ptr byte = addr extValue[0]
+    len = 0
+
+  # Write signature OCTET STRING
+  let retSig = mbedtls_asn1_write_octet_string(
+    addr extPtr, startPtr, unsafeAddr signature[0], signature.len.uint
+  )
+  if retSig < 0:
+    raise newException(ASN1EncodingError, "Failed to write signature OCTET STRING")
+  len += retSig
+
+  # Write publicKey OCTET STRING
+  let retPub = mbedtls_asn1_write_octet_string(
+    addr extPtr, startPtr, unsafeAddr pubKey[0], pubKey.len.uint
+  )
+  if retPub < 0:
+    raise newException(ASN1EncodingError, "Failed to write publicKey OCTET STRING")
+  len += retPub
+
+  # Total length of the SEQUENCE contents
+  let contentLen = retSig + retPub
+  # Write SEQUENCE length
+  let retLen = mbedtls_asn1_write_len(addr extPtr, startPtr, contentLen.uint)
+  if retLen < 0:
+    raise newException(ASN1EncodingError, "Failed to write SEQUENCE length")
+  len += retLen
+
+  # Write SEQUENCE tag
+  let retTag = mbedtls_asn1_write_tag(
+    addr extPtr, startPtr, MBEDTLS_ASN1_CONSTRUCTED or MBEDTLS_ASN1_SEQUENCE
+  )
+  if retTag < 0:
+    raise newException(ASN1EncodingError, "Failed to write SEQUENCE tag")
+  len += retTag
+
+  # Calculate dataOffset based on the accumulated length
+  let dataOffset = extValueSize - len - 1
+
+  # Extract the relevant portion of extValue as a seq[byte]
+  let extValueSeq = toSeq(extValue[dataOffset ..< extValueSize])
+
+  # Return the extension content
+  return extValueSeq
+
+proc makeLibp2pExtension(
+    identityKeypair: KeyPair, certificateKeypair: mbedtls_pk_context
```

Comment:
Maybe it's my C background talking, but I would use a `ptr mbedtls_pk_context` here instead of `mbedtls_pk_context`. It could work in Nim (I'm not sure if it uses a copy, a reference or a pointer for this kind of call), but as mbedtls only use pointer for their context, I would change that for our own function aswell.

Reply:
What's the benefit of it? I don't think this is a petter in Nim, is it?